pinc/DPage.inc

```diff
@@ -716,23 +716,35 @@ function get_normalize_page_text_changes($text, $projectid)
 
     if($invalid_chars)
     {
-        $invalid_char_names = [];
+        // for each element in this array the key will be the character and
+        // the value will be an array of utf8 strings for each component
+        // if not printable the key is U+xxxx codepoint instead.
+        $processed_invalid_chars = [];
         foreach(array_keys($invalid_chars) as $char)
         {
-            $name = IntlChar::charName($char);
-            # if it doesn't have a name it's a control character which can't
-            # be printed anyway, or if it's a soft hyphen
-            if(!$name || utf8_chr_to_hex($char) == 'U+00ad')
+            // find the components; maybe an extended character
+            preg_match_all("/./u", $char, $matches);
+            $char0 = $matches[0][0];
+            // if not defined or is a control character or soft hyphen
+            if(!IntlChar::isdefined($char0) || IntlChar::iscntrl($char0) || ($char0 === "\x{ad}"))
             {
-                $name = utf8_chr_to_hex($char);
+                // put U+xxxx instead of utf8 string
+                $processed_invalid_chars[utf8_chr_to_hex($char0)] = $matches[0];
             }
             else
             {
-                $name = $char;
+                // extended characters will get here unless first component is caught above
+                $processed_invalid_chars[$char] = $matches[0];
             }
-            $invalid_char_names[] = $name;
         }
-        $invalid_char_string = implode(", ", $invalid_char_names);
+        // put the characters in spans with a title
+        $char_spans = [];
+        foreach($processed_invalid_chars as $char => $codepoints)
+        {
+            $title = make_title($codepoints);
+            $char_spans[] = "<span class='warning' title='$title'>$char</span>";
```

Comment:
I don't think these need to be a warning class, they are already in the Warnings column.

Reply:
I do like that these include the title now.

Reply:
True, they don't need to be in the warning class.